ihmeuw/linker
# Feature/mic 4592 container using spark cluster

## Title: Summary, imperative, start upper case, don't end with a period


### Description

- *Category*:  implementation/other
- *JIRA issue*: https://jira.ihme.washington.edu/browse/MIC-4592
- *Research reference*: [see milestone 1.3 and milestone 2.2](https://uwnetid.sharepoint.com/:w:/r/sites/ihme_simulation_science_team/_layouts/15/doc2.aspx?sourcedoc=%7B9805A749-67D7-497B-9EA8-AD8D8C879559%7D&file=Milestones.docx&action=default&mobileredirect=true)

### Changes and notes
 
Matt did the bulk of this work. This PR sorts out how to run a distributed
spark cluster. A followup PR will hook it up to the linker tool.

I squashed all of Matt's work into the first commit to make it easier to see what he did vs
the stuff I did and preserve some of that history since I had to make some significant
changes.

Major change: we decided to skip the pvs-like stuff (milestone 1.x) and jump
right to the dummy/dev stuff (milestone 2.x).

### Verification and Testing

I tested that this is working with the following workflow.

1. Kick off a spark cluster by calling the cli tool and passing in the dummy input file: `linker build-spark-cluster ~/scratch/linker/manual/ ~/scratch/linker/manual/ -e src/linker/specifications/environment_spark.yaml --preserve-logs --input-data /mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/dummy/input_file_1.parquet`
2. Wait for the webui and master urls to be returned to the prompt.
3. Go to the webui and ensure that the nodes are spun up w/ the correct requests.
4. From the unmanaged vm, exec into the implementation container and bind all the things: `singularity shell --containall --no-home --bind /tmp:/tmp --bind ~/scratch/linker/manual/:/results --bind ~/scratch/linker/manual/:/diagnostics --bind /mnt/team/simulation_science/priv/engineering/er_ecosystem/sample_data/dummy/input_file_1.parquet:/input_data/main_input_input_file_1.parquet /mnt/team/simulation_science/priv/engineering/er_ecosystem/containers/python_pyspark_distributed.sif`
5. In the container, export the master url: `export DUMMY_CONTAINER_SPARK_MASTER_URL=<master-url-from-terminal>`
6. In the container, manually run the python script: `/usr/bin/python3 /code/dummy_step.py`
7. Confirm that the output files made their way to the folder I set them to.
8. CRITICAL: Ensure on the webui that the worker nodes were actually used!
